File: 15_caterpillar_method/count_distinct_slices.py
MAX = int(1e9)

def solution(m, a):
  n = len(a)
  cnts = [0] * (m + 1)
  rt = 0
  back = 0
  for front in range(n):
    cnts[a[front]] += 1
    while cnts[a[front]] == 2:
      cnts[a[back]] -= 1
      back += 1
    rt += front - back + 1
    if rt > MAX:
      return MAX
  return rt

File: 15_caterpillar_method/count_distinct_slices.cpp
#include <vector>

using namespace std;

const int MAX = 1e9;

int solution(int m, vector<int> &a) {
  int n = a.size();
  vector<int> cnts(m + 1, 0);
  int rt = 0;
  int back = 0;
  for (int front = 0; front < n; front++) {
    cnts[a[front]]++;
    while (cnts[a[front]] == 2) {
      cnts[a[back]]--;
      back++;
    }
    rt += front - back + 1;
    if (rt > MAX) {
      return MAX;
    }
  }
  return rt;
}
